adapters/intertech/intertech.go

```diff
@@ -0,0 +1,340 @@
+// Request is in shared memory, so we have to make a shallow copy for further modification (imp is already a shallow copy)
+func splitRequestDataByImp(request *openrtb2.BidRequest, imp openrtb2.Imp) openrtb2.BidRequest {
+	requestCopy := *request
+	requestCopy.Imp = []openrtb2.Imp{imp}
+
+	return requestCopy
+}
+
+func getIntertechPlacementId(imp openrtb2.Imp) (*intertechPlacementID, error) {
+	var ext adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+
+	var intertechExt openrtb_ext.ExtImpIntertech
+	if err := json.Unmarshal(ext.Bidder, &intertechExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder: %v", imp.ID, err),
+		}
+	}
+
+	placementID, err := mapExtToPlacementID(intertechExt)
+	if err != nil {
+		return nil, err
+	}
+
+	return placementID, nil
+}
+
+func mapExtToPlacementID(intertechExt openrtb_ext.ExtImpIntertech) (*intertechPlacementID, error) {
+	var placementID intertechPlacementID
+
+	if len(intertechExt.PlacementID) == 0 {
+		placementID.ImpID = strconv.Itoa(int(intertechExt.ImpID))
+		placementID.PageID = strconv.Itoa(int(intertechExt.PageID))
+		return &placementID, nil
+	}
+
+	idParts := strings.Split(intertechExt.PlacementID, "-")
+
+	numericIdParts := []string{}
+
+	for _, idPart := range idParts {
+		if _, err := strconv.Atoi(idPart); err == nil {
+			numericIdParts = append(numericIdParts, idPart)
+		}
+	}
+
+	if len(numericIdParts) < 2 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("invalid placement id, it must contain two parts: %s", intertechExt.PlacementID),
+		}
+	}
+
+	placementID.ImpID = numericIdParts[len(numericIdParts)-1]
+	placementID.PageID = numericIdParts[len(numericIdParts)-2]
+
+	return &placementID, nil
+}
+
+func modifyImp(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		banner, err := modifyBanner(*imp.Banner)
+		if banner != nil {
+			imp.Banner = banner
+		}
+		return err
+	}
+
+	if imp.Native != nil {
+		return nil
+	}
+
+	return &errortypes.BadInput{
+		Message: fmt.Sprintf("Unsupported format. Intertech only supports banner and native types. Ignoring imp id #%s", imp.ID),
+	}
+}
+
+func modifyBanner(banner openrtb2.Banner) (*openrtb2.Banner, error) {
+	format := banner.Format
+
+	if banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0 {
+		if len(format) == 0 {
+			return nil, &errortypes.BadInput{
+				Message: "Invalid size provided for Banner",
+			}
+		}
+
+		firstFormat := format[0]
+		banner.H = &firstFormat.H
+		banner.W = &firstFormat.W
+	}
+
+	return &banner, nil
+}
+
+// "Un-templates" the endpoint by replacing macroses and adding the required query parameters
+func (a *adapter) resolveUrl(placementID intertechPlacementID, referer string, currency string) (string, error) {
+	params := macros.EndpointTemplateParams{PageID: placementID.PageID}
```

Comment:
Per my other comment, you can add an ImpID field in the EndpointTemplateParams struct for this case.

Reply:
Im not sure if I code it correctly

Reply:
No worries. This line should include the imp id to resolve that macro:

`params := macros.EndpointTemplateParams{PageID: placementID.PageID, ImpID: "whatever"}`

Then in add a `"ImpID string" field to the `EndpointTemplateParams` struct here:
https://github.com/prebid/prebid-server/blob/7702a78118cf56ebf2a4e998f4757b01640ef404/macros/macros.go#L9

Reply:
I removed the items you mentioned from the mapExtToPlacmentID func, changed the pattern in intertech.json, and now Im wondering if the "TESTinvalidPlacmentIdMapper" still make sensce. Similary, for unknown-banner, I added the necessary code in 'modifyImps' to handle this case. Im not sure if Im writing tests correctly?
what pattern should I use for placemnt_id should it be (\S+-)?\d+-\d+ or [0-9]+-[0-9]+$ ?